[UnifiedPDF] Selection tracking should be able to reason at word/line granularity

aprotyas opened the pull request:
#### f241c6ee9ea487e2c3f39bcc5c3199406cf81309
<pre>
[UnifiedPDF] Selection tracking should be able to reason at word/line granularity
<a href="https://bugs.webkit.org/show_bug.cgi?id=268616">https://bugs.webkit.org/show_bug.cgi?id=268616</a>
<a href="https://rdar.apple.com/122179178">rdar://122179178</a>

Reviewed by Sammy Gill and Simon Fraser.

This patch adopts new PDFKit API to be able to track selections at a
word or line granularity when the user double or triple clicks on the
page.

* Source/WebKit/Platform/spi/Cocoa/PDFKitSPI.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::continueTrackingSelection):

Canonical link: <a href="https://commits.webkit.org/275085@main">https://commits.webkit.org/275085@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f072e3f482fb72335d3ccab8090a54a18f8b100e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40851 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19864 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43229 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43409 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/36942 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22864 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17195 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33859 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41425 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16811 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35214 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14477 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/14580 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44692 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37065 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36515 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/40253 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/15666 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12864 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/38612 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17285 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9166 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/17336 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/16929 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -2434,21 +2434,35 @@ static FloatRect computeMarqueeSelectionRect(const WebCore::FloatPoint& point1,
+            case SelectionGranularity::Line:
+                return (PDFSelectionGranularity)PDFSelectionGranularityLine;
+            }
+            RELEASE_ASSERT_NOT_REACHED();
```
Is it really necessary to kill the process in release builds?

aprotyas replied:
No, not at all. I'll make this a debug assert instead.

sammygill commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -2434,21 +2434,35 @@ static FloatRect computeMarqueeSelectionRect(const WebCore::FloatPoint& point1,
+                return (PDFSelectionGranularity)PDFSelectionGranularityLine;
+            }
+            RELEASE_ASSERT_NOT_REACHED();
+        };
```
Could we use enumToUnderlyingType since the enum's values in `SelectionGranularity` are in the same order as the `#define`s? Or we need the `#defines`?

aprotyas replied:
So ideally without all the staging code, we'd just map `SelectionGranularity` values to `PDFSelectionGranularity` values, and it's really a hack that `PDFSelectionGranularityCharacter` maps to 0 (which is coincidentally the same underlying type value that `SelectionGranularity::Character` maps to).

I would prefer that we be explicit about this enumeration mapping because really the underlying type value should be opaque and we shouldn't expect the match that we see here... soon enough the `#define`s are going away anyway, so this will be much cleaner 😃

aprotyas committed:
[UnifiedPDF] Selection tracking should be able to reason at word/line granularity
https://bugs.webkit.org/show_bug.cgi?id=268616
rdar://122179178

Reviewed by Sammy Gill and Simon Fraser.

This patch adopts new PDFKit API to be able to track selections at a
word or line granularity when the user double or triple clicks on the
page.

* Source/WebKit/Platform/spi/Cocoa/PDFKitSPI.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::continueTrackingSelection):

Canonical link: https://commits.webkit.org/275085@main
